File: Pokemon-Objc/Pokemon-Objc/PokemonTableViewController.m
//
//  PokemonTableViewController.m
//  Pokemon-Objc
//
//  Created by Chad Parker on 8/14/20.
//  Copyright © 2020 Chad Parker. All rights reserved.
//

#import "PokemonTableViewController.h"
#import "Pokemon_Objc-Swift.h"
#import "LSIPokemon.h"

@interface PokemonTableViewController ()

@property (nonatomic) NSMutableArray<LSIPokemon *> *pokemonArray;

@end

@implementation PokemonTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.pokemonArray = [[NSMutableArray alloc] init];
    
    [PokemonAPI.sharedController fetchAllPokemonWithCompletion:^(NSArray<LSIPokemon *> *pokemonArray, NSError *error) {
        self.pokemonArray = [pokemonArray mutableCopy];
        [self.tableView reloadData];
    }];
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.pokemonArray.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"PokemonCell" forIndexPath:indexPath];
    
    LSIPokemon *pokemon = self.pokemonArray[indexPath.row];
    cell.textLabel.text = pokemon.name;
    
    return cell;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}

@end

File: Pokedex-ObjC/Pokedex-ObjC/View Controlllers/PokemonTableViewController.swift
//
//  PokemonTableViewController.swift
//  Pokedex-ObjC
//
//  Created by Lambda_School_Loaner_263 on 2/4/20.
//  Copyright © 2020 Lambda_School_Loaner_263. All rights reserved.
//

import UIKit

class PokemonTableViewController: UITableViewController {
    
    // -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
    // MARK: - Properties
    var pokemons: [PNCPokemon] = []
    
    
    override func viewDidLoad() {
        super.viewDidLoad()

        loadPokemon()
    }
    
    // -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
    // MARK: - Methods
    private func loadPokemon() {
           
           PokemonAPI.shared.fetchAllPokemon { (pokemons, error) in
               if let error = error {
                   print("Error fetching pokemon in view controller: \(error)")
                   return
               }
               
               guard let pokemons = pokemons else {
                   print("No pokemons was returned")
                   return
               }
               
               self.pokemons.append(contentsOf: pokemons)
               print(pokemons)
               DispatchQueue.main.async {
                   self.tableView.reloadData()
               }
           }
       }

    // MARK: - Table view data source

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        return pokemons.count
    }

    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = tableView.dequeueReusableCell(withIdentifier: "PokemonCell", for: indexPath)
    
       cell.textLabel?.text = pokemons[indexPath.row].name
    
    return cell
    }
    
    override func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        
        if tableView.contentOffset.y >= (tableView.contentSize.height - tableView.frame.size.height) {
            print("end")
            loadPokemon()
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
    
        if segue.identifier == "ShowPokemonDetailSegue" {
            guard let detailVC = segue.destination as? PokemonDetailViewController,
                let indexPath = tableView.indexPathForSelectedRow else { return }
            
            let pokemon = pokemons[indexPath.row]
            detailVC.pokemon = pokemon
        }
     }

}
